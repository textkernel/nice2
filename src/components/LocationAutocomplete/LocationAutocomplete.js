import React from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import { FaMapMarkerAlt } from 'react-icons/fa';
import { Autosuggest, Alert, ListItem, MarkedText } from '../../index';
import useDebounce from '../../hooks/useDebounce';
import POWERED_BY_GOOGLE_ON_WHITE from './images/powered_by_google_on_white.png';
import POWERED_BY_GOOGLE_ON_WHITE_2X from './images/powered_by_google_on_white@2x.png';
import POWERED_BY_GOOGLE_ON_WHITE_3X from './images/powered_by_google_on_white@3x.png';
import styles from './LocationAutocomplete.scss';

const { elem } = bem({
    name: 'LocationAutocomplete',
    classnames: styles,
});

const DEBOUNCE_DELAY = 350;
const ACCEPTABLE_API_STATUSES = ['OK', 'NOT_FOUND', 'ZERO_RESULTS'];

const LocationAutocomplete = props => {
    const [storage] = React.useState({ latestInputValue: '' });
    const [suggestionsList, setSuggestionsList] = React.useState(null);
    const [isLoading, setIsLoading] = React.useState(false);
    const [inputValue, setInputValue] = React.useState('');
    const debouncedInputValue = useDebounce(inputValue, DEBOUNCE_DELAY);

    const {
        isFocused,
        onSelectionChange,
        inputPlaceholder,
        noSuggestionsPlaceholder,
        country,
        placeTypes,
        showCountryInSuggestions,
        onError,
        hidePoweredByGoogleLogo,
        ...rest
    } = props;

    // Suggestion functions
    const resetSuggestionsList = () => setSuggestionsList(null);
    const suggestionToString = suggestion => (suggestion ? suggestion.description : '');

    React.useEffect(() => {
        if (debouncedInputValue) {
            // Putting latest debounced input value to the storage
            storage.latestInputValue = debouncedInputValue;

            const service = new window.google.maps.places.AutocompleteService();

            service.getPlacePredictions(
                {
                    input: debouncedInputValue,
                    types: placeTypes,
                    componentRestrictions: { country },
                },
                (predictions, status) => {
                    // if this function was called with outdated input value, return early
                    if (debouncedInputValue !== storage.latestInputValue) return;

                    if (ACCEPTABLE_API_STATUSES.includes(status)) {
                        setSuggestionsList(predictions);
                    } else {
                        // TODO: check desired behaviour with Carlo
                        // currently the UI will look same as when no suggestions found
                        resetSuggestionsList();
                        if (onError) {
                            onError(status);
                        }
                    }
                    setIsLoading(false);
                }
            );
        } else {
            resetSuggestionsList();
        }
    }, [country, debouncedInputValue, onError, placeTypes, storage.latestInputValue]);

    if (!(window.google && window.google.maps && window.google.maps.places)) {
        // TODO: clarify with Carlo how to handle errors
        return (
            <Alert context="bad" title="No API found">
                Google Maps Places APi was not found on the page. Before using this component, make
                sure to load the places API
            </Alert>
        );
    }

    const handleInputValueChange = value => {
        setInputValue(value);
        if (value) {
            setIsLoading(true);
        } else {
            setIsLoading(false);
            resetSuggestionsList();
        }
    };

    const handleSelection = value => {
        resetSuggestionsList();
        setInputValue('');
        onSelectionChange(value);
    };

    // eslint-disable-next-line react/display-name
    const renderListPoweredByGoogle = ({ listInputValue, getItemProps, highlightedIndex }) => {
        const elems = suggestionsList.map((item, index) => (
            <ListItem
                key={suggestionToString(item)}
                {...getItemProps({
                    item,
                    index,
                    isHighlighted: highlightedIndex === index,
                    highlightContext: 'brand',
                })}
            >
                <MarkedText marker={listInputValue} inline>
                    {showCountryInSuggestions
                        ? suggestionToString(item)
                        : item.structured_formatting.main_text}
                </MarkedText>
            </ListItem>
        ));

        if (!hidePoweredByGoogleLogo) {
            elems.unshift(
                <img
                    key="powered by google logo"
                    {...elem('poweredByGoogleImage', props)}
                    src={POWERED_BY_GOOGLE_ON_WHITE}
                    srcSet={`${POWERED_BY_GOOGLE_ON_WHITE}, ${POWERED_BY_GOOGLE_ON_WHITE_2X} 2x, ${POWERED_BY_GOOGLE_ON_WHITE_3X} 3x`}
                    alt="Powered by Google"
                    data-list-exception
                />
            );
        }

        return elems;
    };

    return (
        <Autosuggest
            getSuggestions={suggestionsList}
            suggestionToString={suggestionToString}
            isLoading={isLoading}
            isFocused={isFocused}
            inputPlaceholder={inputPlaceholder}
            noSuggestionsPlaceholder={noSuggestionsPlaceholder}
            listRenderer={renderListPoweredByGoogle}
            onBlur={resetSuggestionsList}
            onInputValueChange={handleInputValueChange}
            onSelectionChange={handleSelection}
            iconNode={<FaMapMarkerAlt {...elem('icon', props)} />}
            {...rest}
        />
    );
};

LocationAutocomplete.displayName = 'LocationAutocomplete';

LocationAutocomplete.propTypes = {
    /** to be shown in the input field when no value is typed */
    inputPlaceholder: PropTypes.string.isRequired,
    /** to be shown when no suggestions are available */
    noSuggestionsPlaceholder: PropTypes.string.isRequired,
    /** trigger of the initial focus of the input field */
    isFocused: PropTypes.bool,
    /** callback to be called with selected value.
     * Value is of type AutocompletePrediction: https://developers.google.com/maps/documentation/javascript/reference/places-autocomplete-service#AutocompletePrediction
     */
    onSelectionChange: PropTypes.func.isRequired,
    /** restrict predictions to country/countries.
     * For details see: https://developers.google.com/maps/documentation/javascript/reference/places-autocomplete-service#ComponentRestrictions
     */
    country: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
    /** type of locations that should be searched for.
     * For details see: https://developers.google.com/maps/documentation/javascript/reference/places-autocomplete-service#AutocompletionRequest.types
     */
    placeTypes: PropTypes.arrayOf(PropTypes.string),
    /** show state and country in suggestions list */
    showCountryInSuggestions: PropTypes.bool,
    /** function to be executed if error occurs while fetching suggestions */
    onError: PropTypes.func,
    /** To hide powered by google logo. For legal reasons only set it to true if Google map is displayed on the same screen as this component! */
    hidePoweredByGoogleLogo: PropTypes.bool,
};

LocationAutocomplete.defaultProps = {
    country: null,
    placeTypes: ['(regions)'],
    isFocused: false,
    showCountryInSuggestions: false,
    onError: null,
    hidePoweredByGoogleLogo: false,
};

export default LocationAutocomplete;
