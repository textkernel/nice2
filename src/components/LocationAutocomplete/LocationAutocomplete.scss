.LocationAutocomplete {
    &__poweredByGoogleImage {
        position: absolute;
        right: var(--spacing-4x);
        bottom: var(--spacing-4x);
    }

    &__icon {
        color: var(--color-muted);
    }
}
