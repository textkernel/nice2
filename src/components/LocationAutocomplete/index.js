export {
    default as LocationAutocompleteWithGoogleLoader,
} from './LocationAutocompleteWithGoogleLoader';
export { default as LocationAutocomplete } from './LocationAutocomplete';
