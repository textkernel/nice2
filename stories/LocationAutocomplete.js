import React from 'react';
import { storiesOf } from '@storybook/react';
import { LocationAutocomplete } from '@textkernel/oneui';

storiesOf('Organisms|LocationAutocomplete', module).add('LocationAutocomplete', () => (
    <LocationAutocomplete apiKey="" />
));
