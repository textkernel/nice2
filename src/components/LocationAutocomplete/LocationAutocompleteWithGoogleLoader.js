import * as React from 'react';
import PropTypes from 'prop-types';
import { LoadScriptNext } from '@react-google-maps/api';
import LoadingSpinner from '../LoadingSpinner';
import LocationAutocomplete from './LocationAutocomplete';

// for details see: https://developers.google.com/maps/documentation/javascript/libraries
const GOOGLE_API_LIBRARIES = ['places'];

const LocationAutocompleteWithGoogleLoader = props => {
    const { apiKey, language, region, additionalGoogleProps, ...rest } = props;

    return (
        <LoadScriptNext
            googleMapsApiKey={apiKey}
            language={language}
            region={region}
            loadingElement={<LoadingSpinner centerIn="parent" />}
            libraries={GOOGLE_API_LIBRARIES}
            {...additionalGoogleProps}
        >
            <LocationAutocomplete {...rest} />
        </LoadScriptNext>
    );
};

LocationAutocompleteWithGoogleLoader.displayName = 'LocationAutocompleteWithGoogleLoader';

LocationAutocompleteWithGoogleLoader.propTypes = {
    /** Google API key */
    apiKey: PropTypes.string.isRequired,
    /**
     * The language code to be used for the map (e.g en). By default the users browser language will be used
     * For available values see: https://developers.google.com/maps/faq#languagesupport
     */
    language: PropTypes.string,
    /** Regional setting for the map. By default Google uses US.
     * For details see: https://developers.google.com/maps/documentation/javascript/localization#Region
     */
    region: PropTypes.string,
    /** other props to pass to the google loader. For details see: https://react-google-maps-api-docs.netlify.com/#loadscriptnext */
    additionalGoogleProps: PropTypes.object, // eslint-disable-line react/forbid-prop-types
};

LocationAutocompleteWithGoogleLoader.defaultProps = {
    additionalGoogleProps: {},
    language: undefined,
    region: undefined,
};

export default LocationAutocompleteWithGoogleLoader;
