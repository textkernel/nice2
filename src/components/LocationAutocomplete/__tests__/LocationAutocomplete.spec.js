import React from 'react';
import toJson from 'enzyme-to-json';
import LocationAutocomplete from '../LocationAutocomplete';

describe('<LocationAutocomplete/> that loads google api and renders a LocationAutocomplete', () => {
    it('should render with default props', () => {
        const wrapper = mount(
            <LocationAutocomplete
                apiKey="someKey"
                inputPlaceholder="Location..."
                noSuggestionsPlaceholder="No suggestions..."
                onSelectionChange={() => null}
            />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
