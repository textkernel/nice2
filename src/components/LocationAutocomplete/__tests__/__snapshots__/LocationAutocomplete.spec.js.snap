// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationAutocomplete/> that loads google api and renders a LocationAutocomplete should render with default props 1`] = `
<LocationAutocomplete
  additionalGoogleProps={Object {}}
  apiKey="someKey"
  inputPlaceholder="Location..."
  noSuggestionsPlaceholder="No suggestions..."
  onSelectionChange={[Function]}
>
  <LoadScriptNext
    googleMapsApiKey="someKey"
    libraries={
      Array [
        "places",
      ]
    }
    loadingElement={
      <LoadingSpinner
        centerIn="parent"
        context="brand"
        hidden={false}
        size={null}
      />
    }
  >
    <LoadingSpinner
      centerIn="parent"
      context="brand"
      hidden={false}
      size={null}
    >
      <div
        aria-busy={true}
        aria-hidden={false}
        className="LoadingSpinner LoadingSpinner--centerIn_parent"
        role="status"
      >
        <svg
          className="LoadingSpinner__svg"
          style={null}
          viewBox={
            Array [
              0,
              0,
              44,
              44,
            ]
          }
        >
          <circle
            className="LoadingSpinner__path LoadingSpinner__path--context_brand"
            cx="22"
            cy="22"
            fill="none"
            r="20"
            strokeWidth="4"
          />
        </svg>
      </div>
    </LoadingSpinner>
  </LoadScriptNext>
</LocationAutocomplete>
`;
